sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala

```diff
@@ -366,9 +364,8 @@ case class FileSourceScanExec(
   @transient
   private lazy val pushedDownFilters = {
     val supportNestedPredicatePushdown = DataSourceUtils.supportNestedPredicatePushdown(relation)
-    // TODO: should be able to push filters containing metadata columns down to skip files
     dataFilters.filterNot(_.references.exists {
```

Comment:
I think it's better to remove the metadata col filter in `FileSourceStrategy`, as that's where we make the metadata col filter invalid: we remove the struct-type metadata col and convert its fields to attributes.

Reply:
agreed, it is better to remove in `FileSourceStrategy`, but we actually need to use the intact `dataFilters` here in `FileSourceScanExec` to filter files to read (i.e. `listFiles(..., dataFilters)`). 

added some comments here, do you think it is OK for now?